Recursive group membership found on an item

See https://community.openhab.org/t/recursive-group-membership-found/135192

I noticed this on 3.3.0-SNAPSHOT - Build #2852 

Error log:
```
2022-04-16 22:10:20.165 [ERROR] [rest.core.item.EnrichedItemDTOMapper] - Recursive group membership found: BackPorch_Light2 is both, a direct or indirect parent and a child of gOutsideLights.
2022-04-16 22:10:20.166 [ERROR] [rest.core.item.EnrichedItemDTOMapper] - Recursive group membership found: BackPorch_Light1 is both, a direct or indirect parent and a child of gOutsideLights.
2022-04-16 22:10:20.166 [ERROR] [rest.core.item.EnrichedItemDTOMapper] - Recursive group membership found: BackYard_Light is both, a direct or indirect parent and a child of gOutsideLights.
2022-04-16 22:10:20.166 [ERROR] [rest.core.item.EnrichedItemDTOMapper] - Recursive group membership found: BackPorch_Light3 is both, a direct or indirect parent and a child of gOutsideLights.
```

Items:

```
Group gLights
Group gOutsideLights  (gLights)

Group:Switch:OR(ON,OFF) gBackYardLights "Backyard Lights" (gOutsideLights) {ga="Light", synonyms="Back Porch Lights, Back Yard Lights"}

Group:Switch:OR(ON,OFF) gBackPorch_Switch2 


Switch             BackPorch_Light2                         "Back Porch Light 2"                                                   <switch>           (gOutsideLights, gBackPorch_Switch2, gBackYardLights, gBackPorch)                                                                      ["Lightbulb"]                      { channel="mqtt:topic:backporch-switch2:power1", autoupdate="false", ga="Light", alexa="Light" }
Switch             BackPorch_Light1                         "Back Porch Light 1"                                                   <switch>           (gBackPorch, gOutsideLights, gBackYardLights, gBackPorch_Switch2)                                                                      ["Lightbulb"]                      { channel="mqtt:topic:backporch-switch2:power2", autoupdate="false", ga="Light", alexa="Light" }
Switch             BackYard_Light                           "Back Porch Light 4"                                                   <switch>           (gBackPorch, gOutsideLights, gBackYardLights, gBackPorch_Switch2)                                                                      ["Lightbulb"]                      { channel="mqtt:topic:backporch-switch2:power3", autoupdate="false", ga="Light" }
Switch             BackPorch_Light3                         "Back Porch Light 3"                                                   <switch>           (gBackPorch, gOutsideLights, gBackYardLights, gBackPorch_Switch2)                                                                      ["Lightbulb"]                      { channel="mqtt:topic:backporch-switch2:power4", autoupdate="false", ga="Light", alexa="Light" }
```

--- a/bundles/org.openhab.core.io.rest.core/pom.xml
+++ b/bundles/org.openhab.core.io.rest.core/pom.xml
@@ -41,6 +41,12 @@
       <version>${project.version}</version>
       <scope>test</scope>
     </dependency>
+    <dependency>
+      <groupId>ch.qos.logback</groupId>
+      <artifactId>logback-classic</artifactId>
+      <version>1.2.3</version>
+      <scope>test</scope>
+    </dependency>
   </dependencies>
 
 </project>

--- a/bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/item/EnrichedItemDTOMapper.java
+++ b/bundles/org.openhab.core.io.rest.core/src/main/java/org/openhab/core/io/rest/core/item/EnrichedItemDTOMapper.java
@@ -64,16 +64,19 @@ public class EnrichedItemDTOMapper {
         return map(item, itemDTO, drillDown, itemFilter, uriBuilder, locale, new ArrayList<>());
     }
 
-    private static EnrichedItemDTO map(Item item, boolean drillDown, @Nullable Predicate<Item> itemFilter,
+    private static EnrichedItemDTO mapRecursive(Item item, @Nullable Predicate<Item> itemFilter,
             @Nullable UriBuilder uriBuilder, @Nullable Locale locale, List<Item> parents) {
         ItemDTO itemDTO = ItemDTOMapper.map(item);
-        return map(item, itemDTO, drillDown, itemFilter, uriBuilder, locale, parents);
+        return map(item, itemDTO, true, itemFilter, uriBuilder, locale, parents);
     }
 
     private static EnrichedItemDTO map(Item item, ItemDTO itemDTO, boolean drillDown,
             @Nullable Predicate<Item> itemFilter, @Nullable UriBuilder uriBuilder, @Nullable Locale locale,
             List<Item> parents) {
-        parents.add(item);
+        if (item instanceof GroupItem) {
+            // only add as parent item if it is a group, otherwise duplicate memberships trigger false warnings
+            parents.add(item);
+        }
         String state = item.getState().toFullString();
         String transformedState = considerTransformation(state, item, locale);
         if (transformedState != null && transformedState.equals(state)) {
@@ -88,7 +91,7 @@ public class EnrichedItemDTOMapper {
             link = null;
         }
 
-        EnrichedItemDTO enrichedItemDTO = null;
+        EnrichedItemDTO enrichedItemDTO;
 
         if (item instanceof GroupItem) {
             GroupItem groupItem = (GroupItem) item;
@@ -101,10 +104,10 @@ public class EnrichedItemDTOMapper {
                                 "Recursive group membership found: {} is both, a direct or indirect parent and a child of {}.",
                                 member.getName(), groupItem.getName());
                     } else if (itemFilter == null || itemFilter.test(member)) {
-                        members.add(map(member, drillDown, itemFilter, uriBuilder, locale, parents));
+                        members.add(mapRecursive(member, itemFilter, uriBuilder, locale, parents));
                     }
                 }
-                memberDTOs = members.toArray(new EnrichedItemDTO[members.size()]);
+                memberDTOs = members.toArray(new EnrichedItemDTO[0]);
             } else {
                 memberDTOs = new EnrichedItemDTO[0];
             }

--- a/bundles/org.openhab.core.io.rest.core/src/test/java/org/openhab/core/io/rest/core/item/EnrichedItemDTOMapperTest.java
+++ b/bundles/org.openhab.core.io.rest.core/src/test/java/org/openhab/core/io/rest/core/item/EnrichedItemDTOMapperTest.java
@@ -17,17 +17,43 @@ import static org.hamcrest.MatcherAssert.assertThat;
 import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.openhab.core.items.GenericItem;
 import org.openhab.core.items.GroupItem;
 import org.openhab.core.library.CoreItemFactory;
+import org.openhab.core.library.items.NumberItem;
 import org.openhab.core.test.java.JavaTest;
+import org.slf4j.LoggerFactory;
+
+import ch.qos.logback.classic.Level;
+import ch.qos.logback.classic.Logger;
+import ch.qos.logback.classic.spi.ILoggingEvent;
+import ch.qos.logback.core.read.ListAppender;
 
 /**
  * @author Kai Kreuzer - Initial contribution
  */
 @NonNullByDefault
 public class EnrichedItemDTOMapperTest extends JavaTest {
+
+    private @NonNullByDefault({}) ListAppender<ILoggingEvent> listAppender;
+
+    @BeforeEach
+    public void setup() {
+        // add logging interceptor
+        Logger logger = (Logger) LoggerFactory.getLogger(EnrichedItemDTOMapper.class);
+        logger.setLevel(Level.ERROR);
+        listAppender = new ListAppender<>();
+        listAppender.start();
+        logger.addAppender(listAppender);
+    }
+
+    @AfterEach
+    public void tearDown() {
+        listAppender.stop();
+    }
 
     @Test
     public void testFiltering() {
@@ -84,6 +110,12 @@ public class EnrichedItemDTOMapperTest extends JavaTest {
         groupItem2.addMember(groupItem1);
 
         assertDoesNotThrow(() -> EnrichedItemDTOMapper.map(groupItem1, true, null, null, null));
+
+        assertThat(listAppender.list.size(), is(1));
+        ILoggingEvent loggingEvent = listAppender.list.get(0);
+        assertThat(loggingEvent.getLevel(), is(Level.ERROR));
+        assertThat(loggingEvent.getFormattedMessage(), is(
+                "Recursive group membership found: group1 is both, a direct or indirect parent and a child of group2."));
     }
 
     @Test
@@ -97,5 +129,26 @@ public class EnrichedItemDTOMapperTest extends JavaTest {
         groupItem3.addMember(groupItem1);
 
         assertDoesNotThrow(() -> EnrichedItemDTOMapper.map(groupItem1, true, null, null, null));
+
+        assertThat(listAppender.list.size(), is(1));
+        ILoggingEvent loggingEvent = listAppender.list.get(0);
+        assertThat(loggingEvent.getLevel(), is(Level.ERROR));
+        assertThat(loggingEvent.getFormattedMessage(), is(
+                "Recursive group membership found: group1 is both, a direct or indirect parent and a child of group3."));
+    }
+
+    @Test
+    public void testDuplicateMembershipOfPlainItemsDoesNotTriggerWarning() {
+        GroupItem groupItem1 = new GroupItem("group1");
+        GroupItem groupItem2 = new GroupItem("group2");
+        NumberItem numberItem = new NumberItem("number");
+
+        groupItem1.addMember(groupItem2);
+        groupItem1.addMember(numberItem);
+        groupItem2.addMember(numberItem);
+
+        EnrichedItemDTOMapper.map(groupItem1, true, null, null, null);
+
+        assertThat(listAppender.list.size(), is(0));
     }
 }

